RFC: support for HSM private keys in TLS handshake

OYTIS opened the pull request:
The use-case is briefly described here: https://github.com/nodejs/node/issues/28921. I want to use a private key managed by an OpenSSL engine to establish a TLS connection.

Another option, `sigalgs`, is useful, among the other things, to adapt to the limitations of the hardware (e.g. if it doesn't support RSAPSS). I can make it a separate PR if it makes more sense.

Usage example:

```
tls = require('tls')
tls_ctx = tls.createSecureContext( {'privateKeyEngine':'mycoolengine',
                                    'engineKey':'myenginekey',
                                    'ca':ca_cert,
                                    'cert':client_cert,
                                    'sigalgs' : 'RSA+SHA384:RSA+SHA256'})
socket = tls.connect({'host':'host.domain.tld','port':8080,'secureContext':tls_ctx})
```

OYTIS commented:
Example usage:

```
tls = require('tls')
tls_ctx = tls.createSecureContext({'privateKeyEngine':'mycoolengine','engineKey':'myenginekey', 'ca':ca_cert, 'cert':client_cert, 'sigalgs' : 'RSA+SHA384:RSA+SHA256'})
socket = tls.connect({'host':'host.domain.tld','port':8080,'secureContext':tls_ctx})
```

jasnell commented:
Needs documentation and a unit test

Trott commented:
@nodejs/crypto

tniessen commented on lib/_tls_common.js:
```diff
@@ -153,6 +153,38 @@ exports.createSecureContext = function createSecureContext(options) {
   }
 
+  const engineKey = options.engineKey;
+  const privateKeyEngine = options.privateKeyEngine;
```
Nit: Can be shortened to

```javascript
const { engineKey, privateKeyEngine } = options;
```

(And probably include `sigalgs` as well if you like.)

tniessen commented on lib/_tls_common.js:
```diff
@@ -153,6 +153,38 @@ exports.createSecureContext = function createSecureContext(options) {
+  const privateKeyEngine = options.privateKeyEngine;
+  if (engineKey && privateKeyEngine) {
+    if ((typeof options.engineKey === 'string') &&
+        (typeof options.privateKeyEngine === 'string')) {
```
Nit: Unnecessary parantheses around simple equality checks.

tniessen commented on lib/_tls_common.js:
```diff
@@ -153,6 +153,38 @@ exports.createSecureContext = function createSecureContext(options) {
+  const engineKey = options.engineKey;
+  const privateKeyEngine = options.privateKeyEngine;
+  if (engineKey && privateKeyEngine) {
+    if ((typeof options.engineKey === 'string') &&
```
Don't extract the property from the `options` object twice: Use `engineKey` here, not `options.engineKey`. (In JavaScript, there is no guarantee that `options.engineKey === options.engineKey`.)

tniessen commented on lib/_tls_common.js:
```diff
@@ -153,6 +153,38 @@ exports.createSecureContext = function createSecureContext(options) {
+        c.context.setEngineKey(engineKey, privateKeyEngine);
+      else
+        throw new ERR_CRYPTO_CUSTOM_ENGINE_NOT_SUPPORTED();
+    } else if (typeof options.engineKey != 'string') {
```
Use `engineKey` instead of `options.engineKey`. Use `!==` instead of `!=`.

tniessen commented on lib/_tls_common.js:
```diff
@@ -153,6 +153,38 @@ exports.createSecureContext = function createSecureContext(options) {
+    } else if (typeof options.engineKey != 'string') {
+      throw new ERR_INVALID_ARG_TYPE('options.engineKey',
+                                     ['string', 'null', 'undefined'],
+                                     options.engineKey);
```
`options.engineKey` → `engineKey`.

tniessen commented on lib/_tls_common.js:
```diff
@@ -153,6 +153,38 @@ exports.createSecureContext = function createSecureContext(options) {
+    }
+  }
+
+  if (options.sigalgs) {
```
Nit: You could extract this option along with the others above using destructuring.

tniessen commented on lib/_tls_common.js:
```diff
@@ -153,6 +153,38 @@ exports.createSecureContext = function createSecureContext(options) {
+    } else {
+      throw new ERR_INVALID_ARG_TYPE('options.privateKeyEngine',
+                                     ['string', 'null', 'undefined'],
+                                     options.privateKeyEngine);
```
`options.privateKeyEngine` → `privateKeyEngine`.

tniessen commented on lib/_tls_common.js:
```diff
@@ -153,6 +153,38 @@ exports.createSecureContext = function createSecureContext(options) {
+    }
+
+    if (c.context.setClientSigalgs)
+      c.context.setClientSigalgs(options.sigalgs);
```
As above, don't extract the property multiple times. Doing that allows to bypass the type validation.

tniessen commented on src/node_crypto.cc:
```diff
@@ -745,6 +747,73 @@ void SecureContext::SetKey(const FunctionCallbackInfo<Value>& args) {
+  unsigned int len = args.Length();
+  if (len != 2) {
+    return THROW_ERR_MISSING_ARGS(env,
+                 "Private key name and engine name are required");
```
If this function is not supposed to be called directly, use `CHECK_EQ(len, 2)` instead. I know that the TLS code base is not a great example, but this kind of input validation should not happen in C++.

tniessen commented on src/node_crypto.cc:
```diff
@@ -745,6 +747,73 @@ void SecureContext::SetKey(const FunctionCallbackInfo<Value>& args) {
+  char errmsg[1024];
+  const node::Utf8Value engine_id(env->isolate(), args[1]);
+  ENGINE* e = LoadEngineById(*engine_id, &errmsg);
+  if (!e) {
```
`e == nullptr`.

tniessen commented on src/node_crypto.cc:
```diff
@@ -745,6 +747,73 @@ void SecureContext::SetKey(const FunctionCallbackInfo<Value>& args) {
+  if (len != 1) {
+    return THROW_ERR_MISSING_ARGS(env,
+                "Colon-separated list of signature algoritms is required");
+  }
```
Same as above: Use `CHECK_EQ(len, 1)` instead if this function is not supposed to be called directly.

tniessen commented on src/node_crypto.cc:
```diff
@@ -745,6 +747,73 @@ void SecureContext::SetKey(const FunctionCallbackInfo<Value>& args) {
+    if (!err) {
+      return env->ThrowError("ENGINE_load_private_key");
+    }
+    return ThrowCryptoError(env, err);
```
I _think_ you can replace this block with

```cpp
if (!key) {
  return ThrowCryptoError(env, ERR_get_error(), "ENGINE_load_private_key")`.
}
```

tniessen commented on src/node_crypto.cc:
```diff
@@ -745,6 +747,73 @@ void SecureContext::SetKey(const FunctionCallbackInfo<Value>& args) {
+
+  int rv = SSL_CTX_use_PrivateKey(sc->ctx_.get(), key.get());
+
+  if (!rv) {
```
`rv == 0`.

tniessen commented on src/node_crypto.cc:
```diff
@@ -745,6 +747,73 @@ void SecureContext::SetKey(const FunctionCallbackInfo<Value>& args) {
+    if (!err)
+      return env->ThrowError("SSL_CTX_use_PrivateKey");
+    return ThrowCryptoError(env, err);
+  }
```
Same as above, I think you can replace the whole block with `ThrowCryptoError`.

tniessen commented on src/node_crypto.cc:
```diff
@@ -745,6 +747,73 @@ void SecureContext::SetKey(const FunctionCallbackInfo<Value>& args) {
+  if (!rv) {
+    unsigned long err = ERR_get_error();  // NOLINT(runtime/int)
+    if (!err)
+      return env->ThrowError("SSL_CTX_set1_sigalgs_list");
```
See above, I think you can use `ThrowCryptoError` instead of this block.

tniessen commented on lib/_tls_common.js:
```diff
@@ -153,6 +153,38 @@ exports.createSecureContext = function createSecureContext(options) {
 
+  const engineKey = options.engineKey;
+  const privateKeyEngine = options.privateKeyEngine;
+  if (engineKey && privateKeyEngine) {
```
Node.js has traditionally been lazy when it comes to input validation, but new APIs should be stricter: Use `!== undefined` or something similar to not ignore inputs such as `engineKey: 0`.

tniessen replied:
Also, I don't know if it would make sense to validate types even if only one option is given? Or to even throw if only one is given? Is there any use case for only specifying one of these options?

OYTIS replied:
> Is there any use case for only specifying one of these options?

@tniessen From the OpenSSL point of view, you can extract multiple keys from one engine, but since we only have one key in the SecureContext, they should probably go together. On the other hand, for client certificate we have a separate option to set the engine, and then to set the certificate, but there is a OpenSSL API call for setting certificate engine, while there is nothing to set "key engine".

tniessen replied:
I agree, but I would prefer to throw when only one of the options is given. (Currently, you are ignoring the options if only one is used.) If there is no legitimate use case for only specifying one of the two options, then it is most likely an error or a mistake, and we shouldn't silently ignore that.

sam-github replied:
I'm not sure this comment was addressed, though the code was outdated, since if only one of engine or keyId is provided, they are  both ignored.

tniessen commented:
I can't comment much on the matter, maybe @bnoordhuis or @sam-github can, I never used OpenSSL engines. These options would be for server keys then, not for client keys?

The other addition, `sigalgs`, is completely unrelated, isn't it? Consider splitting these changes into two PRs if you would like. Maybe we should also consider other names for this option.

OYTIS commented:
@tniessen Thank you for your review, your comments are extremely valuable for a newcomer like myself. I am addressing all of them in the coming version.

> These options would be for server keys then, not for client keys?
These options are for the communication party under control of Node.js. If it's a server (which is a more common scenario with Node.js, as far as I know), then it will be server keys, the example in my first comment is rather a client.

OYTIS commented:
Fixed hopefully all the issues, added documentation and tests. `sigalgs` part was removed, another PR is coming

vsemozhetbyt commented on doc/api/tls.md:
```diff
@@ -1425,6 +1425,11 @@ changes:
     `object.passphrase` is optional. Encrypted keys will be decrypted with
     `object.passphrase` if provided, or `options.passphrase` if it is not.
+  * `privateKeyEngine` {string} Name of an OpenSSL engine to get private key
+    from. Should be use together with `engineKey`, otherwise ignored.
```
Nit:
```suggestion
    from. Should be used together with `engineKey`, otherwise ignored.
```

vsemozhetbyt commented on doc/api/tls.md:
```diff
@@ -1425,6 +1425,11 @@ changes:
+  * `privateKeyEngine` {string} Name of an OpenSSL engine to get private key
+    from. Should be use together with `engineKey`, otherwise ignored.
+  * `engineKey` {string} Identifier of a private key managed by OpenSSL engine.
+    Should be use together with `privateKeyEngine`, otherwise ignored. Overrides
```
Nit:
```suggestion
    Should be used together with `privateKeyEngine`, otherwise ignored. Overrides
```

Trott commented:
> I can't comment much on the matter, maybe @bnoordhuis or @sam-github can, I never used OpenSSL engines.

/ping @nodejs/crypto @rvagg Any opinions on this? If we're definitely not going to merge this, I'd like to close it. But if we might merge it, it would be great to nudge it forward a bit....

rvagg commented on test/parallel/test-tls-keyengine-unsupported.js:
```diff
@@ -0,0 +1,34 @@
+
+binding.SecureContext = function() {
+  const rv = new NativeSecureContext();
+  rv.setEngineKey = undefined;
```
what is this test proving? when in the real world is `setEngineKey()` unavailable and why would `ERR_CRYPTO_CUSTOM_ENGINE_NOT_SUPPORTED` ever be thrown? it seems to me that the path to throwing that error is not possible without being able to modify the binding in this way.

OYTIS replied:
@rvagg Honestly, that was almost literally copied from `test-tls-clientcertengine-unsupported.js`, and I'm not sure what the original purpose was, and if it is equally useful for key engines.

rvagg replied:
wrapping in `OPENSSL_NO_ENGINE` made this test meaningful, if mildly so.

sam-github replied:
I'd be OK with this test being deleted, its testing internals an an obscure linker condition.

rvagg commented:
> If we're definitely not going to merge this

I don't see a reason to reject this out of hand, seems like a reasonable ask if the code can get up to scratch. Solid tests are tricky to do for it though, without building and loading in a custom engine so I'm not sure how to view that problem—maybe in the same way that we view the problem of testing commits that deal with LibreSSL? Caveat emptor, it's expected the committer and users of the code in question do the testing IRL? Is that an OK standard for crypto?

rvagg commented on lib/_tls_common.js:
```diff
@@ -153,6 +153,24 @@ exports.createSecureContext = function createSecureContext(options) {
+  const { engineKey, privateKeyEngine } = options;
+  if (engineKey !== undefined && privateKeyEngine !== undefined) {
+    if (typeof engineKey === 'string' && typeof privateKeyEngine === 'string') {
+      if (c.context.setEngineKey)
```
As per my comment below, I don't see how you get a false here unless you're manually messing with the binding as in the test.

Does `OPENSSL_NO_ENGINE` help us at all? Maybe you can wrap the `"setEngineKey"` in an `ifndef` for that like `setClientCertEngine` is?

OYTIS replied:
> Maybe you can wrap the "setEngineKey" in an ifndef for that like setClientCertEngine is?

Yes, probably I forgot this, thank you, will fix. Can't tell what `OPENSSL_NO_ENGINE` is in general for though.

sam-github replied:
It's describes when OpenSSL is built without engine capability.

OYTIS replied:
Yes, `setEngineKey` is added conditionally now:

https://github.com/nodejs/node/pull/28973/files#diff-801e3948990f4965a8ea4aca4a423864R475

OYTIS commented:
@vsemozhetbyt @rvagg @mhdawson  @jasnell Thank you for your reviews, everything should be fixed now (CI job should be restarted probably, it seems the failure is unrelated to the PR).

I agree that a dummy engine is the way to test this functionality meaningfully. If you decide it's the way to go, I can try to implement that.

mhdawson commented:
I'd agree with @rvagg that it would be better to avoid adding code that we can't test. If you can try to create a test with a dummy engine that would be great.

rvagg commented:
if you can give it a go that would be great @OYTIS, I don't imagine it'll be simple but if you come up with a mechanism to compile and load a custom engine during test time then we could go back and apply it to the custom engine use for certs.

sam-github commented:
We kindof support engines, in that there is code and some docs, but we don't test the support, so its hard to know it works, what it works for, if the docs are correct, etc.

All of which is to say, if someone has the time to add a dummy engine that would be great. Or possibly, if there is a software only engine available on Linux, perhaps the tests could search for it, and use it if available? That would allow testing locally for developers, and allow us to add a container test that uses engines.

OYTIS commented:
Sorry, I'm a bit short on time now, I'll try to finish that shortly.

From what I can see the test in `test/addons/openssl-client-cert-engine` is almost what we need here, just for cert engines.

OYTIS commented:
Again, the CI failure looks irrelevant to the PR, needs restarting.

OYTIS commented:
Oh, it's even crashing on some platforms (such as `linuxone`). Thank you for running that.

Would it be reasonable then to restrict this test to Linux, like `certengine` test is restricted to Mac (just because it's the only platform I have at my disposal)? Or I can set `-fPIC` on other platforms as well, and see if it makes things better.

sam-github commented on src/node_crypto.cc:
```diff
@@ -745,6 +748,42 @@ void SecureContext::SetKey(const FunctionCallbackInfo<Value>& args) {
+    return env->ThrowError(errmsg);
+  }
+
+  if (!ENGINE_init(e)) {
```
> To obtain a functional reference from an existing structural reference, call the ENGINE_init() function... otherwise it will return non-zero to indicate that the ENGINE is now operational and will have allocated a new functional reference to the ENGINE. All functional references are released by calling ENGINE_finish() (which removes the implicit structural reference as well).

Should there be a call to `ENGINE_finish()` somewhere?

OYTIS commented:
@sam-github Thank you for your comments, will do the fixes soon. As for `ENGINE_finish()` it is totally reasonable, but we can't do it the similar way as for cert engines, because as far as I understand the functional reference to the engine should live for the whole lifespan of the key. One way of solving this is to add `ENGINE*` member to the `SecureContext` and finish/free it in the destructor and when a new key is set.

As for the CI, I'm a bit confused to be honest. Travis job fails to find `libopenssl.a` in its place, although in compilation phase it was clearly compiled and put into this very directory. The same seems to be the case for Jenkins Linux job. Probably the safest bet would be to restrict the test for `mac` platform, as it is done for `openssl-client-cert-engine` test. Would that work?

sam-github commented:
restricting to osx is fine for now, we can look later at what the build problem is

OYTIS commented on doc/api/tls.md:
```diff
@@ -1440,6 +1440,11 @@ changes:
+    from. Should be used together with `engineKey`, otherwise ignored.
+  * `engineKey` {string} Identifier of a private key managed by OpenSSL engine.
+    Should be used together with `privateKeyEngine`, otherwise ignored.
+    Overrides the key set with `key`.
```
Would it be a better/cleaner solution to interpret `key` option differently depending on whether `privateKeyEngine` is set (and get rid of `engineKey` alltogether)?

sam-github replied:
I've mixed feelings, that it would integrate better with current APIs is compelling. But, what would this do?

```
key: [
 {pem: '---- BEGIN..'},
  'key-for-sam-in-a-HSM'
],
 keyEngine: '/some/key/hsm.so'
```

Would it do something expected, like choosing the right key for the cert chosen to match the key exchange algorithm?

If not, or if only a HSM key can be used, it might be better to have a different param, like you do now, rather than complicate the docs and meaning of the existing key param.

richardlau commented on test/addons/openssl-key-engine/binding.gyp:
```diff
@@ -0,0 +1,41 @@
+      'includes': ['../common.gypi'],
+      'conditions': [
+        ['OS!="mac" and '
+         'OS!="axi" and '
```
```suggestion
         'OS!="aix" and '
```

richardlau commented on test/addons/openssl-key-engine/binding.gyp:
```diff
@@ -0,0 +1,41 @@
+          'include_dirs': ['../../../deps/openssl/openssl/include'],
+          'link_settings': {
+            'libraries': [
+              '../../../../out/<(obj_dir)/deps/openssl/<(openssl_product)'
```
This isn't going to work on Travis or some of our Linux configurations (e.g. ARM) where compilation of the main `node` binary happens in a different job to running tests (i.e. the intermediate compiled artifacts for `deps` are not guaranteed to be available in the test jobs).

OYTIS commented:
Addressed style issues, made the test osx-only and added cleanup.

sam-github commented on lib/_tls_common.js:
```diff
@@ -153,6 +153,31 @@ exports.createSecureContext = function createSecureContext(options) {
+  const { privateKeyIdentifier, privateKeyEngine } = options;
+  if (privateKeyIdentifier !== undefined && privateKeyEngine !== undefined) {
+    if (key) {
+      // Both cleartext key and engine key can't be set at the same time
```
nit: `key:` isn't necessarily cleartext, ciphertext/encrypted PEM is supported.

sam-github commented on lib/_tls_common.js:
```diff
@@ -153,6 +153,31 @@ exports.createSecureContext = function createSecureContext(options) {
   }
 
+  const { privateKeyIdentifier, privateKeyEngine } = options;
+  if (privateKeyIdentifier !== undefined && privateKeyEngine !== undefined) {
```
Was it your intention that specifying a `privateKeyIdentifier` while forgetting to set the `privateKeyEngine` would be silently ignored? I can see an argument for ignoring the engine if no key identifier is specified, but I'm having more trouble seeing the case for ignoring the key if the engine is not specified.

OYTIS replied:
Yes, that was the idea, but I can throw if that is better.

sam-github commented on lib/_tls_common.js:
```diff
@@ -153,6 +153,31 @@ exports.createSecureContext = function createSecureContext(options) {
+    if (typeof privateKeyIdentifier === 'string' &&
+        typeof privateKeyEngine === 'string') {
+      if (c.context.setEngineKey)
+        c.context.setEngineKey(privateKeyIdentifier, privateKeyEngine);
```
I guess an empty string is allowed, `''`, because some HSMs might just map that to a default key?

OYTIS replied:
Yes, I think empty string for a key identifier should be fine (I haven't seen an engine that would do that in the wild though).

sam-github commented on lib/_tls_common.js:
```diff
@@ -153,6 +153,31 @@ exports.createSecureContext = function createSecureContext(options) {
+        c.context.setEngineKey(privateKeyIdentifier, privateKeyEngine);
+      else
+        throw new ERR_CRYPTO_CUSTOM_ENGINE_NOT_SUPPORTED();
+    } else if (typeof privateKeyIdentifier !== 'string') {
```
This doesn't support `null`, I'm fine with that, but the warning messages below say that `null` is supported type.

OYTIS replied:
Removed `null` from the messages.

sam-github commented on src/node_crypto.cc:
```diff
@@ -745,6 +748,56 @@ void SecureContext::SetKey(const FunctionCallbackInfo<Value>& args) {
+
+  char errmsg[1024];
+  const node::Utf8Value engine_id(env->isolate(), args[1]);
+  std::unique_ptr<ENGINE, std::function<void(ENGINE*)>> e =
```
Is adding into node_crypto.h a `DeleteFnPtr<ENGINE, ENGINE_finish_and_free_fn>` not possible? If it doesn't work, this is OK, I just want to make sure you noticed the existing smart ptr pattern.

OYTIS replied:
Well, there is a slight variation in that I first create a pointer with a `ENGINE_free_fn` as a deleter, and then replace it with `ENGINE_finish_and_free_fn`. I'm not sure if that will work with `DeleteFnPtr`.

sam-github replied:
makes sense

OYTIS commented:
I believe there is no outstanding issues so far.

sam-github commented on doc/api/tls.md:
```diff
@@ -1440,6 +1440,12 @@ changes:
+    from. Should be used together with `privateKeyIdentifier`.
+  * `privateKeyIdentifier` {string} Identifier of a private key managed by
+    an OpenSSL engine. Should be used together with `privateKeyEngine`.
+    Should not be set together with `key`, because both option define a private
```
nit: `s/option/options/`

sam-github commented:
One  tiny spelling correction, but other than that, can land when it passes CI.

OYTIS commented:
Fixed the typo and added [a section](https://github.com/nodejs/node/pull/28973/files#diff-f6e3a86962eaf0897ab59e88b418e64fR1361-R1364) to `tls.createSecureContext()` documentation.

OYTIS commented:
Rebased, probably Jenkins tests also need to be run.

Trott commented:
Is this ready to land? @jasnell @rvagg @tniessen @sam-github @OYTIS

OYTIS commented:
@Trott I've got no amendments from my side.

sam-github commented:
Yes, landable now that CI is green.

Trott commented:
Landed in c2ce8d0

heol-files commented:
Hi! I have a yubikey, and i guess this is the way to go for me when connecting with TLS?
Do you have any example on how to use/write a "privateKeyEngine"?
Or any documentation on how to do/use this?
